### DISCUSSION TITLE
Is there a way to show x axis labels not link to y axis labels? or make the x axis label space evenly

### DISCUSSION THREAD

**Discussion by VortanaSay01** [patrykandpatrick/vico#704] [Questions] [answered]
I use vico 1.14.0. This is a great library and can't wait for v2 become stable.

**Context:**

I have requirement to show only 5 labels in the x axis in the full width graph (not scrollable). To do that, I create a class implement from AxisItemPlacer.Horizontal interface. Then define the step based on the number of label that the chart want to display. This produce desire behavior.

However, I noticed this implementation just show the label and hide other labels. For example, If I have 10 data points and show only 5 x axis labels with step = 1 (first five elements). I can see 8,9,10,11,12 but the x labels doesn't space evenly across the width. These five elements skew to the left (Refer to screenshot below)  
 
<img width="376" alt="label-not-even" src="https://github.com/patrykandpatrick/vico/assets/117096387/278b1822-31ce-4dcb-9878-a015213e33ff">

**Questions**

- Based on example above, is it possible to have 8,9,10,11,12 space evenly across x axis? Could this achievable part of the class implement AxisItemPlacer.Horizontal? 

- Another thought that I am not sure if this can be done is to shows x axis labels that not link to y axis. For example, show 8.5, 11.5, 15.3, 17.4, 11.0 on the x axis. and these values not exists in the list of chart entry.

Thanks.

**Comment by patrickmichalik** [comment, marked as answer]
Hello! Thanks for the kind words. The labels are positioned at their _x_ values, as specified by the `AxisItemPlacer.Horizontal` implementation. This set of _x_ values isn’t evenly distributed across the coordinate plane, which appears to range from _x_ = 7 to _x_ = 14, and so the labels aren’t evenly distributed along the `HorizontalAxis`.

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/0aac7546-3b9f-4cc1-84d7-f86789f48339" width="480" />

Regarding the first question, could you elaborate on what you’re looking to achieve? You could use `AxisValuesOverrider` to change the upper bound of the `Chart`’s _x_ range from 14 to 13, at which point the labels would be evenly distributed along the `HorizontalAxis`. This, however, is unrelated to `AxisItemPlacer.Horizontal`.

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/34a6e840-83bf-4cec-a7c3-e7f5ee2e36aa" width="480" />

On the other hand, without a change to the _x_ range, for labels with these _x_ values to be evenly distributed along the `HorizontalAxis` would be a mathematical contradiction. You can certainly achieve such label positioning for an _x_ range of 7 through 14, but the label _x_ values should then be 8⅙, 9⅓, and so on. You could, as such, use `AxisValueFormatter` to make these labels read “8 May” through “12 May,” but it should be noted that this would be inaccurate.

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/345465f6-92f7-4bf4-8793-b138e3c4fed6" width="480" />

To address your second question, sure, you can add labels for _x_ values that don’t have any corresponding `ChartEntry` instances, as suggested above. Any _x_ value can be included in the list returned by `AxisItemPlacer.Horizontal#getLabelValues`.

**Reply by VortanaSay01** [reply]
Thanks for the prompt response. 

- **For the first question**, I try to achieve equal space between the label in x axis. Similar to this image you posted, but with accurate value.
<img width="745" alt="Screenshot 2024-05-16 at 9 49 50 AM" src="https://github.com/patrykandpatrick/vico/assets/117096387/e98dfc8d-0fb5-4d18-bbeb-cee785d0842e">
I'll try the option you mentioned.



- **For the second issue,** getLabelValues return List<Float>. 

My current implementation for CustomHorizontalAxisItemPlacer.getLabelValues:

```
// get chartValues
val chartValues = context.chartValuesProvider.getChartValues().chartEntryModel.entries.firstOrNull()

// calculate step based on number of maxItem with simple math

// loop from 0 to maxItem - 1 and add value to values list
  for (index in 0 until maxItemCount) {
                val indexElementForStep = (steps * index).roundToInt()
                values.add(chartValues[indexElementForStep].x)
            }
```
As you can see, my implementation has the values list tie to chartValues, but I want to return list contains values not tie to chartValue, space evenly and between the lower and upper bound of chart value x. For example, let's say we have date x = 05/7 to x = 05/20, I want to have control of let's say show 05/8, 05/10, 05/12, 05/13, 05/18 with space evenly on the x axis.

I get confuse because the getLabelValues return List<Float>, is this float refer the index in chart entry or float of certain list that can not related to chart entry?

**Reply by Gowsky** [reply]
@VortanaSay01 
>with accurate value.

What do you mean by that? Mathematically, these are the only values that give you even spacing.

>05/8, 05/10, 05/12, 05/13, 05/18 with space evenly on the x axis.

How is it supposed to work? Mathematically, these values are not evenly spaced. As @patrickmichalik said, you can use evenly spaced x-axis values and format them as you wish. You can display these labels with even spacing, but it doesn't make sense.

>is this float refer the index in chart entry or float of certain list that can not related to chart entry?

It's explained in the documentation. They're x-axis values and you can calculate them however you want. This a **normal coordinate plane**.
